include/async_scope.hpp

```diff
@@ -624,78 +289,69 @@ namespace _P2519::execution {
       };
 
     namespace __impl {
       template <class _SenderId, class _ReceiverId>
         inline void __operation<_SenderId, _ReceiverId>::__start_() noexcept try {
           if (!!__state_) {
+            std::unique_lock __guard{__state_->__mutex_};
             if (__state_->__data_.index() > 0) {
+              __guard.unlock();
               __complete_();
             } else {
-              __state_->__push_back_(this);
+              __state_->__subscribers_.push_back(this);
             }
-          } else {
-            set_stopped((_Receiver&&) __rcvr_);
           }
         } catch(...) {
           set_error((_Receiver&&) __rcvr_, std::current_exception());
         }
     } // namespace __impl
 
-    template <sender _Sender>
-      inline void __future_state<_Sender>::__push_back_(__impl::__subscription* __subscription) {
-        std::unique_lock __lock{__mutex_};
-        if (__head_ == nullptr) {
-          __head_ = __subscription;
-        } else {
-          __tail_->__next_ = __subscription;
-        }
-        __tail_ = __subscription;
-        __subscription->__next_ = nullptr;
-        __cv_.notify_one();
-      }
-
-    template <sender _Sender>
-      inline __impl::__subscription* __future_state<_Sender>::__pop_front_() {
-        std::unique_lock __lock{__mutex_};
-        if (__head_ == nullptr) {
-          return nullptr;
-        }
-        auto* __subscription = __head_;
-        __head_ = __subscription->__next_;
-        if (__head_ == nullptr)
-          __tail_ = nullptr;
-        return __subscription;
-      }
-
     template <class _SenderId>
       class __future {
         using _Sender = __t<_SenderId>;
         friend struct async_scope;
-        template <class _Receiver>
-          using __completions = completion_signatures_of_t<_Sender, env_of_t<_Receiver>>;
-
+      public:
+        ~__future() noexcept {
+          if (__state_ != nullptr) {
+            auto __raw_state = __state_.get();
+            std::unique_lock __guard{__raw_state->__mutex_};
+            __raw_state->__no_future_ = std::move(__state_);
+            __raw_state->__step_from_to_(__guard, __future_state_steps::Future, __future_state_steps::NoFuture);
+          }
+        }
+        __future(__future&&) = default;
+        __future& operator=(__future&&) = default;
+      private:
         explicit __future(std::unique_ptr<__future_state<_Sender>> __state) noexcept
           : __state_(std::move(__state))
-        {}
+        {
+            std::unique_lock __guard{__state_->__mutex_};
+            __state_->__step_from_to_(__guard, __future_state_steps::Created, __future_state_steps::Future);
+        }
 
-        template <class _Receiver>
-            requires receiver_of<_Receiver, __completions<_Receiver>>
+        template <__decays_to<__future> _Self, class _Receiver>
+            // requires receiver_of<_Receiver, __completions<_Receiver>>
```

Comment:
What's the story here? This function really needs to be constrained.

Reply:
fixed